### PR TITLE
DYT-3712: Redact credential URLs from engine/connection log messages

### PR DESCRIPTION
**Pull request by msrdic**
## Summary
- Added `_safe_url_for_log()` helper in `src/khora/storage/_log_safe.py` that strips username/password from connection URLs before they appear in log messages, replacing credentials with `<redacted>`
- Applied the helper across all storage backends that log connection URLs: Neo4j, Memgraph, Neptune, SurrealDB, SQLite+LanceDB relational layer, StorageFactory (shared engine cache), and Weaviate temporal store
- Passwords and usernames embedded in bolt://, postgresql+asyncpg://, and other URL schemes will no longer appear in plaintext in log output
- URLs without credentials are passed through unchanged

## Test plan
- [x] Unit tests pass (`tests/unit/storage/test_log_safe.py` — 6 parametrized cases covering user+password, user-only, no credentials, non-default port, sqlite file paths)
- [x] `make format` — clean
- [x] `make lint` — clean (ruff)
- [x] Full unit test suite passes (3142 tests)

### CODE REVIEW

**Comment by msrdic** [comment]
## Audit Summary

| Audit | Status |
|-------|--------|
| Self-audit (`/audit --auto`) | clean (fixed in 2 cycles) |
| Specialist: `/review:test --auto` | clean (fixed in 2 cycles) |
| Bandit (Python security) | clean (fixed in 2 cycles) |
| Cross-audit (`/codex-audit`) | ran |

**Cycles:** 2/5
**Result:** clean (fixed in 2 cycles)

---

## Fix Cycle 1

### Files changed
```
src/khora/storage/backends/surrealdb/connection.py | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)
```

### Diff
```diff
diff --git a/src/khora/storage/backends/surrealdb/connection.py b/src/khora/storage/backends/surrealdb/connection.py
index 11c4b75..9148590 100644
--- a/src/khora/storage/backends/surrealdb/connection.py
+++ b/src/khora/storage/backends/surrealdb/connection.py
@@ -8,6 +8,8 @@ from typing import Any
 
 from loguru import logger
 
+from ..._log_safe import _safe_url_for_log
+
 # Module-level lock prevents concurrent schema initialization from
 # multiple SurrealDBConnection instances (the StorageCoordinator
 # connects 4 backends in parallel, all sharing the same embedded DB).
@@ -89,7 +91,7 @@ class SurrealDBConnection:
         import surrealdb
 
         endpoint = self._build_endpoint()
-        logger.info(f"Connecting to SurrealDB ({self._mode}): {endpoint}")
+        logger.info(f"Connecting to SurrealDB ({self._mode}): {_safe_url_for_log(endpoint)}")
         # Use getattr to create the client so ty does not infer the union
         # return type of AsyncSurreal() (whose HTTP variant .connect(url)
         # has a required url param that conflicts with the WS variant).
```

### Validation

| Metric | Count |
|--------|-------|
| Fixed (verified) | 0 |
| Fixed (unverified) | 0 |
| Accepted | 0 |
| Skipped | 0 |
| Missing evidence | 0 |
| **Total expected** | **0** |

---

## Audit Outcome: VERIFIED FIXED

All HIGH/MEDIUM findings from earlier cycles were verified fixed by a clean audit in cycle 2.

# Audit Findings — DYT-3712

**Branch:** `ms/DYT-3712-khora-redact-credential-urls-from`  
**Base:** `origin/main`  
**Cycle:** 2 (verification audit)  
**Date:** 2026-05-05

## Verification Status of Cycle-1 Findings

### M-1: SurrealDB remote connection logs raw endpoint URL — RESOLVED ✓

`src/khora/storage/backends/surrealdb/connection.py` now imports `_safe_url_for_log` at line 11 and applies it at line 94:
```python
logger.info(f"Connecting to SurrealDB ({self._mode}): {_safe_url_for_log(endpoint)}")
```
This finding is resolved and is **not carried forward**.

---

## HIGH Findings

*None.*

---

## MEDIUM Findings

*None.*

---

## LOW Findings

### L-1: Private namedtuple API used in `_log_safe.py`
**File:** `src/khora/storage/_log_safe.py:12`  
**Severity:** LOW  
**Status:** Still present

```python
return parsed._replace(netloc=f"<redacted>@{host}").geturl()
```

`_replace` is named with a leading underscore by convention (not an officially documented `urllib.parse.ParseResult` API). Cleaner alternative: `urlunparse((parsed.scheme, f"<redacted>@{host}", parsed.path, parsed.params, parsed.query, parsed.fragment))`.

---

### L-2: Test fixture contains a realistic-looking password string
**File:** `tests/unit/storage/test_log_safe.py:11`  
**Severity:** LOW  
**Status:** Still present

```python
"postgresql+asyncpg://peras:ebesspyg7jl0kwkbglto@pgbouncer.example.net/peras",
```

`ebesspyg7jl0kwkbglto` reads as a randomly-generated token, which is ambiguous for future reviewers. Using an obviously-fake string like `hunter2` or `secretpassword` signals intent more clearly.

---

### L-3: Missing edge-case test: URL-encoded credentials
**File:** `tests/unit/storage/test_log_safe.py`  
**Severity:** LOW  
**Status:** Still present

No test covers percent-encoded credentials (e.g., `postgresql://user%40domain:p%40ssword@host/db`). The behavior is correct but undocumented.

---

### L-4: Missing edge-case test: empty-string input
**File:** `tests/unit/storage/test_log_safe.py`  
**Severity:** LOW  
**Status:** Still present

`_safe_url_for_log("")` is not tested. `urlparse("")` returns an empty `ParseResult` with all fields falsy, so the function returns `""` unchanged — correct but undocumented by the test suite.

---

### L-5: IPv6 host URLs produce malformed log strings
**File:** `src/khora/storage/_log_safe.py:9-11`  
**Severity:** LOW  
**Status:** Still present

```python
host = parsed.hostname or ""
if parsed.port:
    host = f"{host}:{parsed.port}"
return parsed._replace(netloc=f"<redacted>@{host}").geturl()
```

`parsed.hostname` strips IPv6 brackets, so `bolt://user:secret@[::1]:7687` becomes `bolt://<redacted>@::1:7687` (malformed — brackets missing). No credential leakage, but the log string is ambiguous. Fix: wrap the hostname in brackets when it contains `:`.

---

## Assessment

| Severity | Count |
|----------|-------|
| HIGH     | 0     |
| MEDIUM   | 0     |
| LOW      | 5     |

**Overall:** M-1 (the only MEDIUM from cycle 1) is confirmed resolved. All remaining findings are LOW. No HIGH or MEDIUM issues remain.

---

## Security Audit

### HIGH

*(None)*

### MEDIUM

*(None)*

### LOW

- **SEC-001: Query-parameter credentials silently bypass redaction** — `src/khora/storage/_log_safe.py:5-7`. `_safe_url_for_log` guards on `parsed.username` and `parsed.password`, which are populated only from the netloc section of the URL. Credentials embedded as query parameters (e.g., `postgresql://host/db?password=secret`) yield `None` for both fields, so the function short-circuits and returns the raw URL unchanged — leaking the password into the log. None of the current callers pass query-param-credential URLs (SQLAlchemy DSNs use netloc credentials), but this is a silent footgun for any future caller that does. **Suggested fix:** also check `parsed.query` for known credential keys (`password`, `pwd`, `auth`) and redact their values, or document the limitation explicitly in the function's docstring.

### Verdict: PASS

No HIGH or MEDIUM findings. All findings are LOW with no confirmed credential exposure in the current call sites. The changes are net security-positive: six previously unredacted URL log calls now have credentials stripped.

**Comment by msrdic** [comment]
## Audits — Fixed 1 finding, 6 LOWs noted (2 cycles)

Fixed:
- 🟡 M-1: SurrealDB remote connection logs raw endpoint URL — fixed

Noted (advisory):
- 🟢 L-1: Private namedtuple API (`_replace`) used in `_log_safe.py` — noted
- 🟢 L-2: Test fixture uses a realistic-looking password string — noted
- 🟢 L-3: Missing edge-case test for URL-encoded credentials — noted
- 🟢 L-4: Missing edge-case test for empty-string input — noted
- 🟢 L-5: IPv6 host URLs produce malformed log strings after redaction — noted
- 🟢 SEC-001: Query-parameter credentials silently bypass redaction — noted
